### DISCUSSION TITLE
Generic udev rule for ttyUSB

### DISCUSSION THREAD

**Discussion by jrozner** [dorssel/usbipd-win#347] [General]
Typically on linux ttyUSB and similar devices will have their group owner be dialout and then providing group membership to a user they can use the device as non-root. Ubuntu in wsl doesn't do any of that, presumably because USB doesn't exist out of the box. Is there a generic udev rule that exists that would provide a similar experience or do I just need to write one that looks for the vid/pid for the specific device I'm plugging in? Specifically I'm using an ftdi board.

**Comment by CraigHutchinson** [comment]
[Not an official/educated response] For time being I have the notes below, note that you can remove the idVendor and idProduct filter if you desire to map any device.
I do often observe that the rules don't seem to be applied on WSL startup so I have to restart the udev service.

```
# Put this into `/etc/udev/rules.d/usb.rules`
SUBSYSTEM=="tty",MODE="0666",GROUP="dialout"
SUBSYSTEM=="usb",ATTR{idVendor}=="04d8",ATTR{idProduct}=="ec5a",MODE="0666",GROUP="dialout"

# (Re)start UDEV to apply rules
sudo service udev restart
sudo udevadm control --reload
```

**Comment by congModmo** [comment]
I use wsl2 to develop embedded software, Restart udev service every boot is boring, so I do this.

write to /etc/wsl.conf
```
[user]
default=<input your username here>

[boot]
command="service udev restart && udevadm control --restart"
```
and it will make all serial belong to dialout group at startup

**Reply by aroberts109** [reply]
I get an error when trying to restart udev service:
"udev does not support containers, not started"

Also, running  **udevadm test $(udevadm info -q path -n /dev/ttyUSB0)**
I see operation not permitted errors when udev rule tries to set mode and group:

```
handling device node '/dev/ttyUSB0', devnum=c188:0, mode=0660, uid=0, gid=20
set permissions /dev/ttyUSB0, 020660, uid=0, gid=20
setting mode of /dev/ttyUSB0 to 020660 failed: Operation not permitted
setting owner of /dev/ttyUSB0 to uid=0, gid=20 failed: Operation not permitted
```

Running WSL2 
Ubuntu 18.04, 
5.10.102.1-microsoft-standard-WSL2
Windows 10 19044.1826

**Comment by jrozner** [comment]
Just closing the loop on this as I've now found what I think is an ideal answer. Enabling systemd on your wsl2 instances will cause the default udev scripts to run when the device is attached and everything just works as normal.
